src/plugins/hetero/src/sync_infer_request.cpp

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "sync_infer_request.hpp"
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "compiled_model.hpp"
+#include "itt.hpp"
+#include "openvino/core/except.hpp"
+#include "openvino/op/util/variable_context.hpp"
+#include "openvino/runtime/iasync_infer_request.hpp"
+#include "openvino/runtime/ivariable_state.hpp"
+#include "openvino/runtime/profiling_info.hpp"
+#include "openvino/runtime/tensor.hpp"
+#include "openvino/util/common_util.hpp"
+#include "plugin.hpp"
+
+// #include "template/remote_tensor.hpp"
+// #include "variable_state.hpp"
+
+using Time = std::chrono::high_resolution_clock;
+
+ov::hetero::InferRequest::InferRequest(const std::shared_ptr<const ov::hetero::CompiledModel>& compiled_model)
+    : ov::ISyncInferRequest(compiled_model) {
+    int index = 0;
+    for (auto&& subnetwork : compiled_model->m_networks) {
+        InferRequest::SubRequestDesc desc;
+        desc._network = subnetwork._network;
+
+        std::string prof_task_name = compiled_model->m_name + "_Req" + std::to_string(index++);
+        desc._profilingTask = {
+            openvino::itt::handle("Hetero_" + prof_task_name + "_StartPipeline"),
+            openvino::itt::handle("Hetero_" + prof_task_name + "_WaitPipline"),
+        };
+        desc._request = {desc._network->create_infer_request(), desc._network._so};
+        m_infer_requests.push_back(desc);
+    }
+
+    for (size_t i = 0; i < compiled_model->inputs().size(); i++) {
+        const auto& port = compiled_model->inputs()[i];
+        const auto& submodel_idx = compiled_model->m_inputs_to_submodel_inputs[i].first;
+        m_port_to_request_idx_map[port] = submodel_idx;
+    }
+    for (size_t i = 0; i < compiled_model->outputs().size(); i++) {
+        const auto& port = compiled_model->outputs()[i];
+        const auto& submodel_idx = compiled_model->m_outputs_to_submodel_outputs[i].first;
+        m_port_to_request_idx_map[port] = submodel_idx;
+    }
+
+    for (const auto& kvp : compiled_model->m_submodels_output_to_input) {
+        const auto& submodel_idx1 = kvp.first.first;
+        const auto& tensor_idx1 = kvp.first.second;
+        const auto& submodel_idx2 = kvp.second.first;
+        const auto& tensor_idx2 = kvp.second.second;
+
+        const auto& output_port = m_infer_requests[submodel_idx1]._network->outputs()[tensor_idx1];
+        const auto& input_port = m_infer_requests[submodel_idx2]._network->inputs()[tensor_idx2];
+        const auto& tensor = m_infer_requests[submodel_idx1]._request->get_tensor(output_port);
+        m_infer_requests[submodel_idx2]._request->set_tensor(input_port, tensor);
+    }
+}
+
+ov::hetero::InferRequest::~InferRequest() = default;
+
+ov::SoPtr<ov::IAsyncInferRequest> ov::hetero::InferRequest::get_request(const ov::Output<const ov::Node>& port) const {
+    auto check_nodes = [](const ov::Node* node1, const ov::Node* node2) {
```

Comment:
Is it possible to fetch common part into some utilities?
https://github.com/openvinotoolkit/openvino/blob/eb3bb52a08044973f9352c0e1f499191497aba14/src/inference/src/dev/isync_infer_request.cpp#L113

Reply:
I guess it won't be neat enough. The final idea is to fix ports comparison, so such lambdas should gone